src/backend/distributed/master/master_citus_tools.c

```diff
@@ -0,0 +1,538 @@
+					   bool *parallel)
+{
+	ArrayType *nodeNameArrayObject = PG_GETARG_ARRAYTYPE_P(0);
+	ArrayType *nodePortArrayObject = PG_GETARG_ARRAYTYPE_P(1);
+	ArrayType *commandStringArrayObject = PG_GETARG_ARRAYTYPE_P(2);
+	bool parallelExecution = PG_GETARG_BOOL(3);
+	int nodeNameCount = ArrayObjectCount(nodeNameArrayObject);
+	int nodePortCount = ArrayObjectCount(nodePortArrayObject);
+	int commandStringCount = ArrayObjectCount(commandStringArrayObject);
+	Datum *nodeNameDatumArray = DeconstructArrayObject(nodeNameArrayObject);
+	Datum *nodePortDatumArray = DeconstructArrayObject(nodePortArrayObject);
+	Datum *commandStringDatumArray = DeconstructArrayObject(commandStringArrayObject);
+	int index = 0;
+	StringInfo *nodeNames = NULL;
+	int *nodePorts = NULL;
+	StringInfo *commandStrings = NULL;
+
+	if (nodeNameCount != nodePortCount || nodeNameCount != commandStringCount)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_SYNTAX_ERROR),
+				 errmsg("expected same number of node name, port, and query string")));
+	}
+
+	nodeNames = palloc0(nodeNameCount * sizeof(StringInfo));
+	nodePorts = palloc0(nodeNameCount * sizeof(int));
+	commandStrings = palloc0(nodeNameCount * sizeof(StringInfo));
+
+	for (index = 0; index < nodeNameCount; index++)
+	{
+		text *nodeNameText = DatumGetTextP(nodeNameDatumArray[index]);
+		char *nodeName = text_to_cstring(nodeNameText);
+		int32 nodePort = DatumGetInt32(nodePortDatumArray[index]);
+		text *commandText = DatumGetTextP(commandStringDatumArray[index]);
+		char *commandString = text_to_cstring(commandText);
+
+		nodeNames[index] = makeStringInfo();
+		commandStrings[index] = makeStringInfo();
+
+		appendStringInfo(nodeNames[index], "%s", nodeName);
+		nodePorts[index] = nodePort;
+		appendStringInfo(commandStrings[index], "%s", commandString);
+	}
+
+	*nodeNameArray = nodeNames;
+	*nodePortsArray = nodePorts;
+	*commandStringArray = commandStrings;
+	*parallel = parallelExecution;
+
+	return nodeNameCount;
+}
+
+
+/*
+ * ExecuteCommandsInParellelAndStoreResults connects to each node specified in
+ * nodeNameArray and nodePortArray, and executes command in commandStringArray
+ * in parallel fashion. Execution success status and result is reported for
+ * each command in statusArray and resultStringArray. Each array contains
+ * commandCount items.
+ */
+static void
+ExecuteCommandsInParellelAndStoreResults(StringInfo *nodeNameArray, int *nodePortArray,
+										 StringInfo *commandStringArray,
+										 bool *statusArray, StringInfo *resultStringArray,
+										 int commmandCount)
+{
+	int commandIndex = 0;
+	char *nodeUser = CurrentUserName();
+	PGconn **connectionArray = palloc0(commmandCount * sizeof(PGconn *));
+	int completedCount = 0;
+
+	/* establish connections */
+	for (commandIndex = 0; commandIndex < commmandCount; commandIndex++)
+	{
+		char *nodeName = nodeNameArray[commandIndex]->data;
+		int nodePort = nodePortArray[commandIndex];
+		PGconn *connection = ConnectToNode(nodeName, nodePort, nodeUser);
```

Comment:
@lithp 1 - omission of transaction was intentional
@lithp 2 - GetShardConnections assume a shard id is available. However, this UDF is meant to connect to any node specified by (address, port). we do things like vacuum, sleep(), pg_table_size(), create index, ... any weird stuff that is not possible by existing master_XXX commands.

@onderkalaci we will most likely revisit this when we wanted to introduce more safety.